### PR TITLE
A faster radix sort.

### PR DESCRIPTION
**Pull request by WilliamDue**
This pull request contains an approximately twice as fast radix sort for large arrays with the same asymptotics. It is based on a paper [1] where they chunk the input, sort it and then scatter the chunks to the array. It adds all the same implementations as `radix_sort*` but with the name `chunked_radix_sort*`. The implementation also takes the arguments `chunk` size. The `chunk` size is the subarray size and I believe it should be similar to the number of threads in a block, on my RTX 3060 I found that 512 worked well.

```
$ futhark bench --backend=cuda bench.fut 
Compiling bench.fut...
Reporting arithmetic mean runtime of at least 10 runs for each dataset (min 0.5s).
More runs automatically performed for up to 300s to ensure accurate measurement.

bench.fut:radix_sort_i32 (no tuning file):
[100000]i32:          1096μs (95% CI: [    1057.3,     1141.3])
[1000000]i32:         7321μs (95% CI: [    7108.0,     7569.1])
[10000000]i32:       71040μs (95% CI: [   70082.7,    72377.2])
[100000000]i32:     696729μs (95% CI: [  694215.0,   699336.7])

bench.fut:chunked_radix_sort_i32 (no tuning file):
[100000]i32:          1098μs (95% CI: [    1009.1,     1203.1])
[1000000]i32:         5171μs (95% CI: [    4821.5,     5555.5])
[10000000]i32:       42105μs (95% CI: [   41243.1,    42905.5])
[100000000]i32:     373699μs (95% CI: [  371335.1,   376460.7])

bench.fut:radix_sort_i64 (no tuning file):
[100000]i64:          2283μs (95% CI: [    2216.2,     2358.4])
[1000000]i64:        17424μs (95% CI: [   17073.3,    17764.2])
[10000000]i64:      163898μs (95% CI: [  162541.2,   165147.8])
[100000000]i64:    1632887μs (95% CI: [ 1628347.8,  1642846.1])

bench.fut:chunked_radix_sort_i64 (no tuning file):
[100000]i64:          2218μs (95% CI: [    2013.9,     2458.9])
[1000000]i64:        11976μs (95% CI: [   11492.9,    12499.3])
[10000000]i64:       92565μs (95% CI: [   91881.0,    93686.3])
[100000000]i64:     858401μs (95% CI: [  856367.9,   860562.1])
```

Something I am using is `intrinsics.partition` to define a `partition4`, I do not know if it is totally fine to do so. I have also tried using it in the current `radix_sort` which did not lead to a speed up. I also distinctly remember trying to implement this algorithm [1] before without the partition4 and it did not lead to a speed up. So I believe the implementation heavily relies on `partition4`.

[1] N. Satish, M. Harris and M. Garland, "Designing efficient sorting algorithms for manycore GPUs," 2009 IEEE International Symposium on Parallel & Distributed Processing, Rome, Italy, 2009, pp. 1-10, doi: 10.1109/IPDPS.2009.5161005.

### CODE REVIEW

**Comment by WilliamDue** [comment]
I have just seen the `--Werror` flag does not allow it to pass the CI, is there a way to tell the compiler this is safe?
It may be the case that the implementation is not safe but it worked according to my `is_sorted` test with the CUDA and OpenCL backend for inputs with 100 million i64.

**Comment by athas** [comment]
Why are you using intrinsics for this?  Just write your own `partition4`.  The intrinsic one does not have any performance advantages.

**Comment by WilliamDue** [comment]
I thought it was a little more special and would make it faster. I am guessing the last time I tried this it was just some bad code.

**Comment by WilliamDue** [comment]
> Why are you using intrinsics for this? Just write your own `partition4`. The intrinsic one does not have any performance advantages.

It uses `radix_sort_step` instead of `intrinsics.partition` now but I looks like it has gotten a little slower. It might also just be the case that my home computer is not the best place to benchmark it.

```
bench.fut:radix_sort_i32 (no tuning file):
[100000]i32:           998μs (95% CI: [     969.6,     1031.8])
[1000000]i32:         7023μs (95% CI: [    6931.2,     7133.4])
[10000000]i32:       66278μs (95% CI: [   65802.5,    66886.8])
[100000000]i32:     653617μs (95% CI: [  652577.9,   654686.3])

bench.fut:chunked_radix_sort_i32 (no tuning file):
[100000]i32:          1029μs (95% CI: [     945.1,     1144.3])
[1000000]i32:         5380μs (95% CI: [    5092.1,     5718.0])
[10000000]i32:       46157μs (95% CI: [   45464.5,    46696.1])
[100000000]i32:     430894μs (95% CI: [  429865.9,   431583.9])

bench.fut:radix_sort_i64 (no tuning file):
[100000]i64:          2048μs (95% CI: [    2013.5,     2087.0])
[1000000]i64:        15379μs (95% CI: [   15199.4,    15492.3])
[10000000]i64:      134430μs (95% CI: [  133539.3,   135168.5])
[100000000]i64:    1316445μs (95% CI: [ 1315411.5,  1318354.4])

bench.fut:chunked_radix_sort_i64 (no tuning file):
[100000]i64:          1735μs (95% CI: [    1680.6,     1802.3])
[1000000]i64:        10849μs (95% CI: [   10619.5,    11049.7])
[10000000]i64:       91053μs (95% CI: [   90155.4,    91291.6])
[100000000]i64:     894142μs (95% CI: [  893269.9,   894389.5])
```

**Comment by WilliamDue** [comment]
Just realized instead of using `radix_sort_step` it would probably be beneficial to choose a smaller type for index since the chunks should be small.

**Comment by WilliamDue** [comment]
It looks like it got a little faster by using i16 but the speed up is not as much as in my first measurements, I should probably try measuring with intrinsics.partition again.
```
bench.fut:radix_sort_i32 (no tuning file):
[100000]i32:           917μs (95% CI: [     892.2,      946.8])
[1000000]i32:         6667μs (95% CI: [    6530.4,     6818.0])
[10000000]i32:       59013μs (95% CI: [   57999.3,    59281.2])
[100000000]i32:     566207μs (95% CI: [  565555.1,   566918.7])

bench.fut:chunked_radix_sort_i32 (no tuning file):
[100000]i32:           828μs (95% CI: [     797.0,      866.8])
[1000000]i32:         4519μs (95% CI: [    4385.2,     4671.0])
[10000000]i32:       38117μs (95% CI: [   37425.4,    38303.2])
[100000000]i32:     357693μs (95% CI: [  356833.3,   357986.6])

bench.fut:radix_sort_i64 (no tuning file):
[100000]i64:          2044μs (95% CI: [    2009.9,     2082.1])
[1000000]i64:        15494μs (95% CI: [   15305.4,    15603.9])
[10000000]i64:      134028μs (95% CI: [  132993.5,   134311.4])
[100000000]i64:    1315576μs (95% CI: [ 1314212.8,  1317446.8])

bench.fut:chunked_radix_sort_i64 (no tuning file):
[100000]i64:          1746μs (95% CI: [    1685.0,     1820.7])
[1000000]i64:        10644μs (95% CI: [   10401.8,    10864.9])
[10000000]i64:       88450μs (95% CI: [   87544.0,    88693.4])
[100000000]i64:     868439μs (95% CI: [  864827.0,   873986.8])
```

**Comment by WilliamDue** [comment]
The parameter `highest` element is no longer needed.

File: lib/github.com/diku-dk/sorts/radix_sort.fut [hunk 1]
```diff
@@ -109,3 +109,184 @@ def radix_sort_float_by_key [n] 't 'k
...
+local def (////) (a: i64) (b: i64) : i64 =
+  a / b + i64.bool (a % b != 0)
+
+-- | This implementation of radix sort works on the outside almost like
+-- `radix_sort` but the implementation is based of a design where you
+-- chunk the input into subarrays [1]. This leads to performance gains
+-- if you choose a good `chunk` size based on the GPU thread block.
+-- Using 512 as a `chunk` size leads to about twice the speed on
+-- a RTX 3060 as the normal `radix_sort`. The implementation also
+-- needs a `highest` element which is used for padding when sorting.
```
Comment (athas) [inline lib/github.com/diku-dk/sorts/radix_sort.fut]: `highest` is gone.

File: lib/github.com/diku-dk/sorts/radix_sort.fut [hunk 2]
```diff
@@ -109,3 +109,184 @@ def radix_sort_float_by_key [n] 't 'k
...
+-- The `highest` element needs to be larger or equal to the largest
+-- element in the input array. The sorting algorithm is stable and its
+-- work is *O(k n)* and the span is *O(k log(n))* as in `radix_sort`
```
Comment (athas) [inline lib/github.com/diku-dk/sorts/radix_sort.fut]: What is `k`?

**Review by athas — CHANGES_REQUESTED** [review]
Was the performance unchanged by the removal of `highest`?

File: lib/github.com/diku-dk/sorts/radix_sort.fut (hunk 2 above)
Comment (LiljeDue) [reply lib/github.com/diku-dk/sorts/radix_sort.fut]: I now mention that this is the number of bits in the elements.

**Comment by WilliamDue** [comment]
It did not lead to any worse performance on large arrays. I end up computing a single `radix_sort_step` on the last part of the array that did not fit into a multiple of the chunk size.
```
bench.fut:radix_sort_i32 (no tuning file):
[100000]i32:           914μs (95% CI: [     890.6,      944.1])
[1000000]i32:         6671μs (95% CI: [    6528.1,     6824.5])
[10000000]i32:       59367μs (95% CI: [   58383.9,    59825.0])
[100000000]i32:     568496μs (95% CI: [  567206.7,   571589.2])

bench.fut:chunked_radix_sort_i32 (no tuning file):
[100000]i32:          1159μs (95% CI: [    1118.7,     1208.6])
[1000000]i32:         4769μs (95% CI: [    4633.5,     4920.9])
[10000000]i32:       38327μs (95% CI: [   37686.2,    38614.1])
[100000000]i32:     359250μs (95% CI: [  357806.5,   360831.0])

bench.fut:radix_sort_i64 (no tuning file):
[100000]i64:          2073μs (95% CI: [    2035.2,     2114.6])
[1000000]i64:        15665μs (95% CI: [   15522.8,    15749.1])
[10000000]i64:      137479μs (95% CI: [  135807.6,   140069.3])
[100000000]i64:    1326490μs (95% CI: [ 1321536.1,  1332019.2])

bench.fut:chunked_radix_sort_i64 (no tuning file):
[100000]i64:          2367μs (95% CI: [    2290.4,     2461.4])
[1000000]i64:        11144μs (95% CI: [   10902.6,    11355.0])
[10000000]i64:       89186μs (95% CI: [   88775.6,    89640.1])
[100000000]i64:     866005μs (95% CI: [  862830.8,   873623.5])
```

File: lib/github.com/diku-dk/sorts/radix_sort.fut (hunk 1 above)
Comment (LiljeDue) [reply lib/github.com/diku-dk/sorts/radix_sort.fut]: Oops forgot to remove this.

**Comment by WilliamDue** [comment]
I read it once more through and realized I did not formulate myself correctly.

File: lib/github.com/diku-dk/sorts/radix_sort.fut [hunk 3]
```diff
@@ -109,3 +109,182 @@ def radix_sort_float_by_key [n] 't 'k
...
+  let (is, elems) =
+    map (
+      \i ->
+        let elem = ys[i]
+        let bin = get_bin get_bit digit_n elem
+        let new_offset = trans_hist_scan[bin][i / m]
+        let old_offset = hist_scan[i / m][bin]
+        let idx = (i - old_offset) + new_offset
+        in (idx, elem)
+    ) (iota (n * m + r))
```
Comment (athas) [inline lib/github.com/diku-dk/sorts/radix_sort.fut]: Use `tabulate` instead of mapping over `iota`.

**Review by athas — CHANGES_REQUESTED** [review]

File: lib/github.com/diku-dk/sorts/radix_sort.fut (hunk 3 above)
Comment (LiljeDue) [reply lib/github.com/diku-dk/sorts/radix_sort.fut]: This should be done now.

**Pull request merged**